Add recipe for dired-duplicates

hjudt opened the pull request:
### Brief summary of what the package does

This Emacs package helps to find duplicate files on local and remote filesystems, showing the results in a Dired buffer. It is similar to the fdupes command-line utility but written in Emacs Lisp and should also work on every remote filesystem that TRAMP supports and where executable commands can be called remotely (the checksumming is still done using an external checksum program for performance reasons).

### Direct link to the package repository

https://codeberg.org/hjudt/find-duplicates

### Your association with the package

I am both the author and maintainer of this package.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm with `x`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  Quick first pass --

`find-duplicates.el` with checkdoc 0.6.2:
```
find-duplicates.el:63: First line is not a complete sentence
find-duplicates.el:73: First line is not a complete sentence
find-duplicates.el:81: First line is not a complete sentence
find-duplicates.el:90: First line is not a complete sentence
find-duplicates.el:108: First line is not a complete sentence
find-duplicates.el:121: First line is not a complete sentence
find-duplicates.el:149: First line is not a complete sentence
find-duplicates.el:182: First line is not a complete sentence
find-duplicates.el:198: First line is not a complete sentence
find-duplicates.el:213: First line is not a complete sentence
```
One reason for these is to make the M-x apropos output not end in the middle of a sentence but it requires some careful rewriting.  E.g.
```elisp
  "Path and name of the separator file used for making search
results easier to discern.  It will be created immediately before
and deleted as soon as possible after the search operation
finishes."
```
could be written
```elisp
  "Path and name of the separator file.
This file is used for making search results easier to discern.
It will be created immediately before and deleted as soon as
possible after the search operation finishes."
```

`find-duplicates.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- find-duplicates.el#L154: There is no need to quote lambdas (neither #' nor ')
- find-duplicates.el#L62: Use `(temporary-file-directory)` instead of /tmp in code; you may also want to see `(make-temp-file)` instead of using `(make-string 40 ?-)`
```

Lastly (and not a big deal) the `error` in `find-duplicates-checksum-file` looks like it should possibly be a `user-error` instead.

hjudt commented:
Thank you for your review and your suggestions, I hope I have fixed all of them correctly, at least flymake doesn't report anything anymore.

hjudt commented:
I have one additional questions: Currently, the package depends on Emacs 28.1. What is the recommended way to make it compatible with earlier versions of Emacs? For example, `make-empty-file` is not available in older versions. How would I best handle this? Check feature with featurep and copy&paste Emacs code?

riscy commented:
> What is the recommended way to make it compatible with earlier versions of Emacs?

Some folks add a package-requires on [compat](https://elpa.gnu.org/packages/compat.html) if you can find the missing function implemented in there.

> Check feature with featurep and copy&paste Emacs code?

This is what some people do as well.  I'd make an effort to avoid creating definitions/aliases outside of your own namespace though, since it may have bad interactions with e.g. compat above.

hjudt commented:
Ok, thanks for clarification. I'll try to address this in a future development.

If I ever want to push this to ELPA, what would I have to do here on MELPA? I'd have no problem with passing ownership to FSF, but I am not sure this package is so useful for other people, and also that it qualifies for ELPA.

hjudt commented:
For now, I have lowered emacs requirements to 27.1 by dropping ensure-list.

riscy commented:
> If I ever want to push this to ELPA, what would I have to do here on MELPA?

You could de-index it from MELPA if you want but there are some packages that exist in both repositories.  They do this because they find it easier to distribute quick updates on MELPA.

In the past we've gently discouraged listing packages in both places to reduce potential confusion or desynchronization.

ELPA requires assigning copyright to the FSF; there's also non-gnu ELPA if you prefer: https://elpa.nongnu.org/

hjudt commented:
@riscy thank you for your explanation. So I could assign copyright to the FSF anyway if I wanted to, I guess it only gets trickier with additional contributors because they would have to give their consent too. This is no problem at the moment ;-)

Is there anything else I need to do to get this package on MELPA, or is everything ok now and I will just have to wait?

riscy commented:
Nope, I think that puts all the regular logistics behind us.  I've got a couple more questions though (albeit non-blocking):

- Have you thought about putting this package under a `dired-` namespace, since that seems to be a central aspect?  E.g. `dired-find-duplicates` or `dired-duplicates` or similar
- I looked around for similar functionality and I found that built into dired is `dired-compare-directories`.  It lets you specify a PREDICATE for comparing two files.  And via TRAMP of course you can specify a remote directory.  This is a bit different than what you've implemented but I mention it because (a) it may be a more general framework you could consider adopting and (b) it seems to have established some norms around making comparisons between directories, without relying on e.g. a separator dummy file.  Not a blocker, like I said, but something I thought I'd bring up.

hjudt commented:
Actually, I have searched for a better name because there already exists find-dupes-dired that uses the fdupes command line tool. Your suggestion of using dired-duplicates or dired-find-duplicates (I like the former more since is shorter) seems a lot nicer though, don't know why this didn't come to my mind.

I chose those separator dummy files (you can turn them off by the way) because I could not make dired output empty lines. If you know of some way to achieve this, that would be great and I could get rid of those dummy files. I need to be able to group the results because some files could have the same size but still be different, and that should stand out. My primary use case was finding duplicate files to delete them.

I've been using find-dupes-dired before, but it does not group the results. fdupes does this by printing empty lines, but it is not very convenient to use. Since I also needed to find duplicate remote files too and because fdupes does not do this, I decided to wield my own solution.

Note one can actually compare many directories using my package, not only one or two, and really find all files that are exactly the same, no matter where they reside. From a quick look at dired-compare-directories, I do not think this framework will scale in that regard, at least I cannot see how this should be implemented to be useful from a user perspective.

hjudt commented:
Shall I first rename the project to dired-duplicates before it is accepted to MELPA? I'd rather do this before and add a new pull request.

riscy commented:
Re: the renaming, that makes sense to me (you could do it in this PR too).

I think you're right about dired-compare-directories not scaling the same way.

hjudt commented:
Ok, I have renamed my repository on codeberg to [dired-duplicates](https://codeberg.org/hjudt/dired-duplicates) and changed all names from find-duplicates to dired-duplicates. From a quick test, it seems to work fine so far, I hope I have not introduced any bugs by doing this. I have not renamed this pull request branch because I do not know whether that might not cause troubles...

I have also renamed the interactive function from `find-duplicates-dired` to `dired-duplicates`.

hjudt commented:
Sigh. It is really impossible to rename my branch for the pull request...

riscy commented:
Not to worry, this is fine.  This look good to merge, but as one parting thought I think you could also add spacing to a dired-mode buffer fairly trivially with e.g. `(let ((inhibit-read-only t)) (insert "\n"))` -- although this might create interactions I haven't thought of, so I'd test carefully.

Merging!

hjudt commented:
Great! I will certainly try your suggestion, as I personally find that separator file approach ugly and it is only a hack. Thanks a lot for your helpful assistance with all the issues and getting this to MELPA.
